File: 2022/day06-1.hs
import Data.List (findIndex, zip4)
import Data.Maybe (fromJust)

solution :: String -> Int
solution content = 4 + (fromJust . findIndex (not . repeating) $ zip4 content (drop 1 content) (drop 2 content) (drop 3 content))

repeating :: (Eq a) => (a, a, a, a) -> Bool
repeating (x, y, z, t) = x == y || x == z || x == t || y == z || y == t || z == t

main :: IO ()
main = do readFile "day06.dat" >>= print . solution

File: 2022/day06-1.ts
function solution(content: string) {
  return (
    4 +
    content
      .split("")
      .map(
        (_char, index, array) =>
          !repeating(
            array[index],
            array[index + 1],
            array[index + 2],
            array[index + 3]
          )
      )
      .findIndex((v) => v)
  );
}

function repeating(a: string, b: string, c: string, d: string): boolean {
  return a === b || a === c || a === d || b === c || b === d || c == d;
}

Deno.readTextFile("day06.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
